File: phantom_core/phantom_core/trust_store.py
"""
Phantom Trust Store — §3 / §5 Append-Only Trust Ledger.

Records every trust-level transition for every worker.
Implements TOFU (Trust On First Use) key management and
key-change detection per the Corrected Architecture Design.
"""

import fcntl
import json
import logging
import os
import threading
import time
from dataclasses import asdict, dataclass
from enum import Enum
from pathlib import Path
from typing import List, Optional

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Event types and trust levels
# ---------------------------------------------------------------------------


class TrustEventType(str, Enum):
    FIRST_SEEN = "first_seen"
    SIGNATURE_VALID = "signature_valid"
    KEY_CHANGED = "key_changed"
    SIGNATURE_INVALID = "signature_invalid"


class TrustLevel(str, Enum):
    UNVERIFIED = "unverified"
    SIG_VALID = "sig_valid"
    APPROVED = "approved"
    REGISTERED = "registered"
    REVOKED = "revoked"


# ---------------------------------------------------------------------------
# TrustRecord — immutable log entry
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class TrustRecord:
    """Immutable record written at every trust-level transition."""

    worker_id: str
    public_key: str
    event_type: str  # TrustEventType value
    trust_level: str  # TrustLevel value
    timestamp: float
    reason: str = ""

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, d: dict) -> "TrustRecord":
        return cls(
            worker_id=d["worker_id"],
            public_key=d["public_key"],
            event_type=d["event_type"],
            trust_level=d["trust_level"],
            timestamp=d["timestamp"],
            reason=d.get("reason", ""),
        )


# ---------------------------------------------------------------------------
# TrustStore — append-only ledger
# ---------------------------------------------------------------------------


class TrustStore:
    """Persistent, append-only trust ledger local to the controller.

    Storage: ``<state_dir>/trust_store.jsonl`` (one JSON record per line).
    Thread-safe via a reentrant lock.
    """

    def __init__(self, state_dir: str):
        self._dir = Path(state_dir).resolve()
        if not self._dir.is_absolute():
            raise ValueError(f"state_dir must resolve to an absolute path: {state_dir}")
        self._dir.mkdir(parents=True, exist_ok=True)
        self._path = self._dir / "trust_store.jsonl"
        self._lock = threading.Lock()
        # In-memory index: worker_id -> list of records (append-only)
        self._records: dict[str, List[TrustRecord]] = {}
        self._load()

    # ---- persistence -------------------------------------------------

    def _load(self) -> None:
        if not self._path.exists():
            return
        try:
            with open(self._path, "r", encoding="utf-8") as f:
                for line_num, line in enumerate(f, 1):
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        rec = TrustRecord.from_dict(json.loads(line))
                        self._records.setdefault(rec.worker_id, []).append(rec)
                    except (json.JSONDecodeError, KeyError, TypeError) as exc:
                        logger.warning(
                            "TrustStore: skipping corrupt record on line %d: %s",
                            line_num,
                            type(exc).__name__,
                        )
        except OSError as exc:
            logger.warning("TrustStore load error: %s", type(exc).__name__)

    def _append_to_file(self, record: TrustRecord) -> None:
        try:
            with open(self._path, "a", encoding="utf-8") as f:
                fcntl.flock(f.fileno(), fcntl.LOCK_EX)
                try:
                    f.write(json.dumps(record.to_dict(), sort_keys=True) + "\n")
                    f.flush()
                    os.fsync(f.fileno())
                finally:
                    fcntl.flock(f.fileno(), fcntl.LOCK_UN)
        except OSError as exc:
            logger.error("TrustStore write error: %s", type(exc).__name__)

    # ---- public API --------------------------------------------------

    def write_record(self, record: TrustRecord) -> None:
        """Append a TrustRecord. Records are never modified or deleted."""
        with self._lock:
            self._records.setdefault(record.worker_id, []).append(record)
            self._append_to_file(record)

    def get_current_level(self, worker_id: str) -> Optional[str]:
        """Return the trust level of the most recent record, or None."""
        with self._lock:
            recs = self._records.get(worker_id)
            if not recs:
                return None
            return recs[-1].trust_level

    def get_current_key(self, worker_id: str) -> Optional[str]:
        """Return the most recently recorded public key for a worker."""
        with self._lock:
            recs = self._records.get(worker_id)
            if not recs:
                return None
            return recs[-1].public_key

    def get_history(self, worker_id: str) -> List[TrustRecord]:
        """Return the full trust history for a worker (oldest first)."""
        with self._lock:
            return list(self._records.get(worker_id, []))

    def get_all_workers(self) -> List[str]:
        """Return all worker IDs that have at least one record."""
        with self._lock:
            return list(self._records.keys())

    # ---- high-level operations --------------------------------------

    def record_verification(
        self, worker_id: str, public_key: str, signature_valid: bool
    ) -> TrustRecord:
        """Record a signature verification result.

        Implements TOFU key-change detection:
        - First contact → first_seen + unverified (then optionally sig_valid)
        - Key change   → key_changed + unverified
        - Valid sig     → signature_valid + sig_valid
        - Invalid sig   → signature_invalid + unverified
        """
        now = time.time()
        existing_key = self.get_current_key(worker_id)

        # Key-change detection
        if existing_key is not None and public_key != existing_key:
            change_rec = TrustRecord(
                worker_id=worker_id,
                public_key=public_key,
                event_type=TrustEventType.KEY_CHANGED.value,
                trust_level=TrustLevel.UNVERIFIED.value,
                timestamp=now,
                reason="key_change_detected",
            )
            self.write_record(change_rec)
            logger.warning("Key change detected for worker %s", worker_id)
            # After key change, still record the sig result below

        # First-seen
        if existing_key is None:
            first_rec = TrustRecord(
                worker_id=worker_id,
                public_key=public_key,
                event_type=TrustEventType.FIRST_SEEN.value,
                trust_level=TrustLevel.UNVERIFIED.value,
                timestamp=now,
                reason="first_contact",
            )
            self.write_record(first_rec)

        # Signature result
        if signature_valid:
            rec = TrustRecord(
                worker_id=worker_id,
                public_key=public_key,
                event_type=TrustEventType.SIGNATURE_VALID.value,
                trust_level=TrustLevel.SIG_VALID.value,
                timestamp=now,
                reason="signature_verified",
            )
        else:
            rec = TrustRecord(
                worker_id=worker_id,
                public_key=public_key,
                event_type=TrustEventType.SIGNATURE_INVALID.value,
                trust_level=TrustLevel.UNVERIFIED.value,
                timestamp=now,
                reason="signature_verification_failed",
            )

        self.write_record(rec)
        return rec

    def approve_worker(self, worker_id: str) -> Optional[TrustRecord]:
        """Promote a worker to approved (user decision)."""
        current = self.get_current_level(worker_id)
        key = self.get_current_key(worker_id) or ""
        if current is None:
            return None
        rec = TrustRecord(
            worker_id=worker_id,
            public_key=key,
            event_type="user_approved",
            trust_level=TrustLevel.APPROVED.value,
            timestamp=time.time(),
            reason="user_approved_worker",
        )
        self.write_record(rec)
        return rec

    def approve_worker_with_key(
        self, worker_id: str, public_key: str
    ) -> TrustRecord:
        """§5 — Record user approval for a worker (possibly first contact).

        Used when the deployment ceremony approves a worker. Creates TrustRecord(approved)
        so the registration endpoint will accept the worker. For new workers (no prior
        record), this establishes the approved trust level.
        """
        key = public_key or ""
        rec = TrustRecord(
            worker_id=worker_id,
            public_key=key,
            event_type="user_approved",
            trust_level=TrustLevel.APPROVED.value,
            timestamp=time.time(),
            reason="user_approved_via_ceremony",
        )
        self.write_record(rec)
        return rec

    def record_registration(self, worker_id: str) -> TrustRecord:
        """§5 — Record that a worker was successfully registered."""
        key = self.get_current_key(worker_id) or ""
        rec = TrustRecord(
            worker_id=worker_id,
            public_key=key,
            event_type="worker_registered",
            trust_level=TrustLevel.REGISTERED.value,
            timestamp=time.time(),
            reason="registration_complete",
        )
        self.write_record(rec)
        return rec

    def revoke_worker(self, worker_id: str) -> Optional[TrustRecord]:
        """Revoke trust for a worker."""
        key = self.get_current_key(worker_id) or ""
        rec = TrustRecord(
            worker_id=worker_id,
            public_key=key,
            event_type="user_revoked",
            trust_level=TrustLevel.REVOKED.value,
            timestamp=time.time(),
            reason="user_revoked_worker",
        )
        self.write_record(rec)
        return rec

File: phantom_app/src-tauri/src/backend/trust_store.rs
//! Phantom §3/§5 — Append-only TrustStore for worker identity tracking.
//!
//! Records every trust-level transition per worker.  Implements TOFU
//! (Trust On First Use) key management and key-change detection.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs::{self, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::path::PathBuf;
use std::sync::Mutex;

// ---------------------------------------------------------------------------
// Event types / trust levels
// ---------------------------------------------------------------------------

/// Event types that trigger a TrustRecord.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum TrustEventType {
    #[serde(rename = "first_seen")]
    FirstSeen,
    #[serde(rename = "signature_valid")]
    SignatureValid,
    #[serde(rename = "key_changed")]
    KeyChanged,
    #[serde(rename = "signature_invalid")]
    SignatureInvalid,
}

/// Current trust level of a worker.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum TrustLevel {
    #[serde(rename = "unverified")]
    Unverified,
    #[serde(rename = "sig_valid")]
    SigValid,
    #[serde(rename = "approved")]
    Approved,
    #[serde(rename = "registered")]
    Registered,
    #[serde(rename = "revoked")]
    Revoked,
}

// ---------------------------------------------------------------------------
// TrustRecord
// ---------------------------------------------------------------------------

/// Immutable record written at every trust-level transition.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrustRecord {
    pub worker_id: String,
    pub public_key: String,
    pub event_type: TrustEventType,
    pub trust_level: TrustLevel,
    pub timestamp: f64,
    #[serde(default)]
    pub reason: String,
}

// ---------------------------------------------------------------------------
// TrustStore
// ---------------------------------------------------------------------------

/// Persistent, append-only trust ledger.
///
/// Storage: `<state_dir>/trust_store.jsonl` (one JSON record per line).
pub struct TrustStore {
    path: PathBuf,
    records: Mutex<HashMap<String, Vec<TrustRecord>>>,
}

impl TrustStore {
    /// Create or open a TrustStore in the given state directory.
    pub fn new(state_dir: &std::path::Path) -> Self {
        let _ = fs::create_dir_all(state_dir);
        let path = state_dir.join("trust_store.jsonl");
        let store = Self {
            path,
            records: Mutex::new(HashMap::new()),
        };
        store.load();
        store
    }

    /// Load records from disk into memory.
    fn load(&self) {
        if !self.path.exists() {
            return;
        }
        let file = match fs::File::open(&self.path) {
            Ok(f) => f,
            Err(e) => {
                log::warn!("TrustStore load error: {e}");
                return;
            }
        };
        let reader = BufReader::new(file);
        let mut map = self.records.lock().unwrap();
        for line in reader.lines() {
            let line = match line {
                Ok(l) => l,
                Err(_) => continue,
            };
            let trimmed = line.trim();
            if trimmed.is_empty() {
                continue;
            }
            if let Ok(rec) = serde_json::from_str::<TrustRecord>(trimmed) {
                map.entry(rec.worker_id.clone())
                    .or_default()
                    .push(rec);
            }
        }
    }

    /// Append a TrustRecord (never modify or delete).
    pub fn write_record(&self, record: &TrustRecord) {
        // In-memory
        {
            let mut map = self.records.lock().unwrap();
            map.entry(record.worker_id.clone())
                .or_default()
                .push(record.clone());
        }
        // Persist
        if let Ok(mut f) = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)
        {
            if let Ok(json) = serde_json::to_string(record) {
                let _ = writeln!(f, "{json}");
            }
        }
    }

    /// Current trust level for a worker (most recent record).
    pub fn get_current_level(&self, worker_id: &str) -> Option<TrustLevel> {
        let map = self.records.lock().unwrap();
        map.get(worker_id)
            .and_then(|recs| recs.last())
            .map(|r| r.trust_level.clone())
    }

    /// Most recently recorded public key for a worker.
    pub fn get_current_key(&self, worker_id: &str) -> Option<String> {
        let map = self.records.lock().unwrap();
        map.get(worker_id)
            .and_then(|recs| recs.last())
            .map(|r| r.public_key.clone())
    }

    /// Full trust history for a worker.
    pub fn get_history(&self, worker_id: &str) -> Vec<TrustRecord> {
        let map = self.records.lock().unwrap();
        map.get(worker_id).cloned().unwrap_or_default()
    }

    /// Record a signature verification result with TOFU key-change detection.
    pub fn record_verification(
        &self,
        worker_id: &str,
        public_key: &str,
        signature_valid: bool,
    ) {
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);

        let existing_key = self.get_current_key(worker_id);

        // Key-change detection
        if let Some(ref old_key) = existing_key {
            if old_key != public_key {
                self.write_record(&TrustRecord {
                    worker_id: worker_id.to_string(),
                    public_key: public_key.to_string(),
                    event_type: TrustEventType::KeyChanged,
                    trust_level: TrustLevel::Unverified,
                    timestamp: now,
                    reason: "key_change_detected".to_string(),
                });
                log::warn!("Key change detected for worker {worker_id}");
            }
        }

        // First-seen
        if existing_key.is_none() {
            self.write_record(&TrustRecord {
                worker_id: worker_id.to_string(),
                public_key: public_key.to_string(),
                event_type: TrustEventType::FirstSeen,
                trust_level: TrustLevel::Unverified,
                timestamp: now,
                reason: "first_contact".to_string(),
            });
        }

        // Signature result
        if signature_valid {
            self.write_record(&TrustRecord {
                worker_id: worker_id.to_string(),
                public_key: public_key.to_string(),
                event_type: TrustEventType::SignatureValid,
                trust_level: TrustLevel::SigValid,
                timestamp: now,
                reason: "signature_verified".to_string(),
            });
        } else {
            self.write_record(&TrustRecord {
                worker_id: worker_id.to_string(),
                public_key: public_key.to_string(),
                event_type: TrustEventType::SignatureInvalid,
                trust_level: TrustLevel::Unverified,
                timestamp: now,
                reason: "signature_verification_failed".to_string(),
            });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    fn tmp_store(name: &str) -> TrustStore {
        let dir = std::env::temp_dir().join("phantom_test_trust").join(name);
        let _ = fs::remove_dir_all(&dir);
        TrustStore::new(&dir)
    }

    #[test]
    fn append_only_produces_multiple_records() {
        let store = tmp_store("append_only");
        store.record_verification("w1", "key_a", true);
        store.record_verification("w1", "key_a", true);
        let history = store.get_history("w1");
        // first_seen + sig_valid + sig_valid (second call only adds sig_valid)
        assert!(history.len() >= 3);
    }

    #[test]
    fn current_level_returns_most_recent() {
        let store = tmp_store("current_level");
        store.record_verification("w1", "key_a", true);
        assert_eq!(
            store.get_current_level("w1"),
            Some(TrustLevel::SigValid)
        );
        store.record_verification("w1", "key_a", false);
        assert_eq!(
            store.get_current_level("w1"),
            Some(TrustLevel::Unverified)
        );
    }

    #[test]
    fn key_change_detection() {
        let store = tmp_store("key_change");
        store.record_verification("w1", "key_a", true);
        store.record_verification("w1", "key_b", true);
        let history = store.get_history("w1");
        let key_changed = history
            .iter()
            .any(|r| r.event_type == TrustEventType::KeyChanged);
        assert!(key_changed, "Expected a key_changed event");
    }

    #[test]
    fn unknown_worker_returns_none() {
        let store = tmp_store("unknown");
        assert_eq!(store.get_current_level("unknown"), None);
        assert_eq!(store.get_current_key("unknown"), None);
    }
}
